operator/pkg/service/envoy_listener.go

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2019-2020 Authors of Cilium
+
+package service
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type serviceAddedEvent struct {
+	service *slim_corev1.Service
+}
+
+type serviceUpdatedEvent struct {
+	oldService *slim_corev1.Service
+	newService *slim_corev1.Service
+}
+
+type serviceDeletedEvent struct {
+	service *slim_corev1.Service
+}
+
+type EnvoyLBController struct {
+	envoyConfigManager *envoyConfigManager
+
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	serviceInformer cache.Controller
+	serviceStore    cache.Store
+}
+
+// NewEnvoyLBController returns a new EnvoyLBController
+func NewEnvoyLBController() (*EnvoyLBController, error) {
+	manager := &EnvoyLBController{
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+	}
+	manager.serviceStore, manager.serviceInformer = informer.NewInformer(
```

Comment:
I'm not super familiar with the operator structure yet, but can this use a shared informer?

Reply:
The current operator structure doesn't have the most modern way of handling K8s watches. Right now, we have central watchers (single Informer) which then fan out to "subscribers". You can see this in `operator/watchers/k8s_service_sync.go`. If we make this `EnvoyLBController` a shared informer, then we can slowly improve the operator code around K8s watch handling. If that isn't possible, then I'd suggest following the "subscriber" pattern as shown in the file I mentioned earlier. 

We certainly don't want to have two duplicated K8s service watchers running for no reason.

Reply:
Thanks, I have refactored the code to use subscriber pattern, one catch is that I didn't use service cache, as each annotation might need to map to one attribute, but we don't know how many annotations we will support right now. I would anticipate like a bunch (10+)

Reply:
The current code is still creating a new informer so it will duplicate the K8s watching that also happens in `operator/watchers/k8s_service_cache.go`. To prevent this, the informer from `operator/watchers/k8s_service_cache.go` should be used instead and the "subscriber" becomes the "operator/pkg/service/envoy_listener.go".

Reply:
does serviceInitOnce help to make sure we initialise informer once only ?

Reply:
@sayboras Yes